### DISCUSSION TITLE
Loading .env on nitro server startup?

### DISCUSSION THREAD

**Discussion by HendrikJan** [nuxt/nuxt#16648] [Questions]
Hi all,

I'm trying to load the .env file into the server.
Now the relative location of the .env file depends on whether I'm doing `nuxi dev` or starting the server after `nuxi build`.
Therefore I'm doing something this:
```javascript
function loadDotenv() {
    import dotenv from 'dotenv';
    import { dirname } from 'path';
    import { fileURLToPath } from 'url';
    const __dirname = dirname(fileURLToPath(import.meta.url));

    if (__dirname.includes('.output')) {
        // On the server, find .env relative to the current file
        dotenv.config({ url: new URL('../../.env', import.meta.url) });
    } else {
        // In development go for the default
        dotenv.config();
    }
}
```

I need to run this on server start and have no idea where to call this function.
There is a `nitro:init` hook, but how can I call this hook (documentation seems not ready yet)?
Or is there a better way?

**Comment by ruslankonev** [comment]
@danielroe pls, how should we use environment variables in a production environment? What about secret tokens? Don't we have to write them in at the start of the application in production?

**Comment by michael-bouvy** [comment]
@HendrikJan @ruslankonev using [env-cmd](https://www.npmjs.com/package/env-cmd) wrapper may be helpful so variables from `.env` are made available into `process.env`:

```bash
# In server build (available under .output/server)
./node_modules/.bin/env-cmd node index.mjs
```

**Reply by HendrikJan** [reply]
That works the same as the "dotenv" package I am using. The problem remains that we don't know the relative location of the.env file, because the Nuxt server executable is located in a different folder for `nuxi dev` and `nuxi build`.

**Reply by michael-bouvy** [reply]
I may be wrong but why does this method need to be called if env variables are already exported by `env-cmd`?

**Reply by HendrikJan** [reply]
Oh, sorry. I didn't read well what env-cmd does.
You are right, env-cmd might help some people to load a .env file before starting the server.

For me, I really want to load .env from within the server.
I've currently got a solution that works and not looking to change it soon, unless Nuxt comes with a nice solution.

**Reply by HendrikJan** [reply]
And thanks for the suggestion!

**Reply by michael-bouvy** [reply]
You're welcome! Can you tell me more about the solution you went for?

**Reply by HendrikJan** [reply]
Sure.
It's far from ideal and a bit of a complex thing, but as I've got other things to work on we'll keep it this way for the time being.
Also because we are rewriting the app from Express+AngularJs (using .mjs files) to Nuxt+Vue (using .ts files), where part of the old code needs to work together with the new, it gets more complex.
There we go:

I've got a file **`/server/plugins/atServerStart.ts`** that will be loaded at server start
```typescript
import { initConfig as initDatabaseConfig, getConfig } from '../config/database.mjs'
import { initConfig as initEmailConfig } from '../config/email.mjs'
import { initConfig as initSomeOtherConfig } from '../config/someOtherConfig.mjs'
import { initConfig as initSapConfig } from '../config/sap.mjs'
import { updateSequelizeConfig } from '../www/helpers/sequelize.mjs' // for legacy use of Sequelize
import fs from 'fs'

export default defineNitroPlugin((nitroApp) => {

    // 1. Read dotenv
    // Try these absolute paths to find the .env file
    const tryDotenvPaths = [
        '/path/to/dotenv/in/production/server1/.env',
        '/path/to/dotenv/in/production/server2/.env',
        '/path/to/dotenv/in/testserver1/.env',
        '/path/to/dotenv/in/testserver2/.env',
        '/path/to/dotenv/in/docker/development/.env',
        '/path/to/dotenv/in/github-actions/.env',
    ]

    for (let dotenvPath of tryDotenvPaths) {
        if (fs.existsSync(dotenvPath)) {
            // A valid path found -- load .env and leave the loop
            console.log('🚀 (Re)Loading config from .env file', dotenvPath)
            dotenv.config({ override: false, path: dotenvPath })
            break
        }
    }

    // 2. Update all configs with the new .env variables
    initDatabaseConfig()
    initEmailConfig()
    initSapConfig()
    initSomeOtherConfig()
    updateSequelizeConfig()
})
```

The config files (like `/server/config/database.mjs`) start out without config set.
This is needed because the content of these files are compiled into the `server.mjs` file that Nuxt builds.
The code above (running `initDatabaseConfig()`) fills the config.
The file **`/server/config/database.mjs`** contains this:
```javascript
 // start with an empty config
const config = {};
// get the config
export const getConfig = () => config;

// initialize the config
export const initConfig = () => {
    Object.assign(config, {
        database_url: process.env.DATABASE_URL,
        database: process.env.DATABASE_URL.split('/').pop(),
        dialect: 'mysql',
        dialectOptions: {
            decimalNumbers: true,
        },
        logging: false,
        domain: process.env.DB_HOST,
        define: {
            paranoid: false,
            underscored: true,
            freezeTableName: true,
        },
        pool: {
            maxConnections: 10,
            max: 10,
            maxIdleTime: 240000,
            Idle: 240000,
        },
        https: true,
    })
};
```

And then wherever I need to use the config:
```javascript
import { getConfig } from '/server/config/database.mjs';

function SomeFunctionThatNeedsConfig() {
    const config = getConfig();
   
    // code that uses the config
}
```

**Reply by michael-bouvy** [reply]
Thanks. Is this Nitro plugin also triggered when starting production build using `node server.mjs`?

**Reply by HendrikJan** [reply]
Good question, I had to try it to be sure.

In our case the `defineNitroPlugin` function is called both during the build (`nuxi build`) and when the server is run (`node .output/server/index.mjs`).

There is one caveat to be aware of -- the `defineNitroPlugin` is run after `nuxt.config.js` is run, so you cannot use .env variable in `nuxt.config.js` (so not in `runtimeConfig` for instance).

**Reply by michael-bouvy** [reply]
Thanks @HendrikJan, I will try this!

**Reply by michael-bouvy** [reply]
@HendrikJan I just tried but in production build, although the Nitro plugin runs fine, the `NUXT_PUBLIC_*` environment variable is not available under `useRuntimeConfig().public` ; passing the env variable from system works fine, i.e. `NUXT_PUBLIC_MY_VAR=helloworld node index.mjs`.

**Reply by HendrikJan** [reply]
My method only works for variables found in the .env file, and only if you call `getConfig()` when you need config.
So indeed I would not expect `useRuntimeConfig` to work with my solution. My solution probably only works well for me and is kind of ugly.
Therefore I'm still waiting for a solution from Nuxt.

**Reply by ines-b** [reply]
Hi, have you had any update?
